src/wpool/gun_server.erl

```diff
@@ -0,0 +1,82 @@
+-module(gun_server).
+
+-behaviour(gen_server).
+
+%% API
+-export([start_link/2]).
+
+%% gen_server callbacks
+-export([init/1,
+         handle_call/3,
+         handle_cast/2]).
+
+-define(SERVER, ?MODULE).
+
+-record(state, {host,
+                port,
+                opts}).
+
+-include("mongoose_logger.hrl").
+
+%%%===================================================================
+%%% API
+%%%===================================================================
+
+start_link(Destination, Options) ->
+    gen_server:start_link(?MODULE, {Destination, Options}, []).
+
+%%%===================================================================
+%%% gen_server callbacks
+%%%===================================================================
+
+-spec(init({{iodata(), inet:port_number()}, map()}) ->
+    {ok, State :: #state{}}).
+init({{Host, Port}, Opts}) ->
+    {ok, #state{host = Host, port = Port, opts = Opts}}.
+
+-spec(handle_call({request, iodata(), iodata(), [{binary(), iodata()}], iodata(), pos_integer(), non_neg_integer()},
+                  From :: {pid(), Tag :: term()}, State :: #state{}) ->
+                     {reply, Reply :: term(), NewState :: #state{}}).
+handle_call({request, FullPath, Method, Headers, Query, _Retries, Timeout}, _From, State) ->
```

Comment:
These line lengths hurts my eyes 😱

Maybe the whole request tuple could have a type on its own, something like a
```erlang
-type request() :: {request,
                    iodata(),
                    iodata(),
                    [{binary(), iodata()}],
                    iodata(),
                    pos_integer(),
                    non_neg_integer()}.
```
(but choose some wiser name, or check if this wasn't defined already in some other pool server and you could refactor and reuse or so).

Also, `-spec` doesn't need parenthesis there. In fact I had no idea it was a valid syntax, never saw it before, but in any case for consistency I'd argue in favour of the parenthesis-less version as in
```erlang
-spec handle_call(x, y) -> z.
```
Same applies for several other specs.

PS: and I'm gonna check that parenthesis, I had no idea 🤔

Reply:
For now the name is a `request()`. Gun does not have such type and neither does `mongoose_http_client` - it would not make much sense there as well.